src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+                    blocks.RemoveAt(i);
+                    i--;
+                    count--;
+                }
+            }
+        }
+
+        private void VisitStatement(IOperation operation)
+        {
+            IOperation saveCurrentStatement = _currentStatement;
+            _currentStatement = operation;
+            Debug.Assert(_evalStack.Count == 0);
+
+            // PROTOTYPE(dataflow): Ensure that statement's parent is null at this point.
+            AddStatement(Visit(operation, null));
+            Debug.Assert(_evalStack.Count == 0);
+            _currentStatement = saveCurrentStatement;
+        }
+
+        private BasicBlock CurrentBasicBlock
+        {
+            get
+            {
+                if (_currentBasicBlock == null)
+                {
+                    AppendNewBlock(new BasicBlock(BasicBlockKind.Block));
+                }
+
+                return _currentBasicBlock;
+            }
+        }
+
+        private void AddStatement(IOperation statement)
+        {
+            if (statement == null)
+            {
+                return;
+            }
+
+            // PROTOTYPE(dataflow): Assert that statement's parent is null at this point.
+            CurrentBasicBlock.AddStatement(statement);
+        }
+
+        private void AppendNewBlock(BasicBlock block)
+        {
+            Debug.Assert(block != null);
+            BasicBlock prevBlock = _blocks.Last();
+
+            if (prevBlock.Next == null)
+            {
+                LinkBlocks(prevBlock, block);
+            }
+
+            _blocks.Add(block);
+            _currentBasicBlock = block;
+        }
+
+        private static void LinkBlocks(BasicBlock prevBlock, BasicBlock nextBlock)
+        {
+            Debug.Assert(prevBlock.Next == null);
+            prevBlock.Next = nextBlock;
+            nextBlock.AddPredecessor(prevBlock);
+        }
+
+        public override IOperation VisitBlock(IBlockOperation operation, object argument)
+        {
+            foreach (var statement in operation.Operations)
+            {
+                VisitStatement(statement);
+            }
+
+            return null;
+        }
+
+        public override IOperation VisitConditional(IConditionalOperation operation, object argument)
+        {
+            if (operation == _currentStatement)
+            {
+                if (operation.WhenFalse == null)
+                {
+                    // if (condition) 
+                    //   consequence;  
+                    //
+                    // becomes
+                    //
+                    // GotoIfFalse condition afterif;
+                    // consequence;
+                    // afterif:
+
+                    BasicBlock afterIf = null;
+                    VisitConditionalBranch(operation.Condition, ref afterIf, false);
+                    VisitStatement(operation.WhenTrue);
+                    AppendNewBlock(afterIf);
```

Comment:
`AppendNewBlock` asserts that its input is non-null. I suspect that means that `VisitConditionalBranch` should always set its `ref` parameter. Could it be changed to `out`?

Reply:
> AppendNewBlock asserts that its input is non-null. I suspect that means that VisitConditionalBranch should always set its ref parameter. Could it be changed to out?

No it cannot. Sometimes it is passed in.

---
In reply to: [162164281](https://github.com/dotnet/roslyn/pull/24263#discussion_r162164281) [](ancestors = 162164281)